File: PolymarketBot/src/polymarket_bot/strategies/fair_value.py
"""
Fair Value strategy — trade when the market price diverges from our probability estimate.

This is the most common Polymarket strategy: you form a view on the true probability
of an event (from your model, external data, news, etc.) and buy whichever token the
market is mispricing relative to that estimate.

Usage:
    strategy = FairValueStrategy(fair_prob=0.65, config=RiskConfig(bankroll_usd=500))
    signal   = strategy.evaluate(snapshot)
"""

from __future__ import annotations

from .base import Action, MarketSnapshot, RiskConfig, Strategy, TradeSignal, kelly_size


class FairValueStrategy(Strategy):
    """
    Single-market fair-value strategy.

    Provide your probability estimate for YES resolving (`fair_prob`).
    The strategy buys YES when market price is too low and buys NO
    when market price is too high — subject to a minimum edge threshold.

    Parameters
    ----------
    fair_prob : float
        Your estimated true probability of YES (0.0 – 1.0).
    config    : RiskConfig
        Risk and sizing parameters.
    """

    def __init__(self, fair_prob: float, config: RiskConfig | None = None) -> None:
        super().__init__(config)
        if not 0.0 < fair_prob < 1.0:
            raise ValueError(f"fair_prob must be between 0 and 1, got {fair_prob}")
        self.fair_prob = fair_prob

    def evaluate(self, snapshot: MarketSnapshot) -> TradeSignal:
        ok, reason = self._passes_quality_checks(snapshot)
        if not ok:
            return self._skip(reason)

        yes_price = snapshot.yes_price
        edge_yes  = self.fair_prob - yes_price   # positive → YES is cheap
        edge_no   = yes_price - self.fair_prob   # positive → NO  is cheap

        if edge_yes >= self.config.min_edge:
            size = kelly_size(self.fair_prob, yes_price, Action.BUY_YES, self.config)
            return TradeSignal(
                action=Action.BUY_YES,
                token_id=snapshot.yes_token_id,
                price=snapshot.yes_ask or yes_price,
                size_usd=size,
                edge=edge_yes,
                reason=f"fair={self.fair_prob:.2f} market={yes_price:.2f} edge={edge_yes:.3f}",
            )

        if edge_no >= self.config.min_edge:
            no_price = snapshot.no_price
            size = kelly_size(self.fair_prob, yes_price, Action.BUY_NO, self.config)
            return TradeSignal(
                action=Action.BUY_NO,
                token_id=snapshot.no_token_id,
                price=snapshot.no_ask or no_price,
                size_usd=size,
                edge=edge_no,
                reason=f"fair={self.fair_prob:.2f} market={yes_price:.2f} edge={edge_no:.3f}",
            )

        return self._skip(
            f"edge too small: yes_edge={edge_yes:.3f} no_edge={edge_no:.3f} min={self.config.min_edge:.3f}"
        )

File: PolymarketBot/functions/src/strategies/fairValue.ts
import { Action, MarketSnapshot, RiskConfig, TradeSignal } from "./types";
import { Strategy }  from "./strategy";
import { kellySize } from "./kellySize";

/**
 * Trade when the market price diverges from our probability estimate.
 *
 * Buys YES when market is underpriced, buys NO when YES is overpriced,
 * subject to a minimum edge threshold.
 */
export class FairValueStrategy extends Strategy {
  private readonly fairProb: number;

  constructor(fairProb: number, config?: Partial<RiskConfig>) {
    super(config);
    if (fairProb <= 0 || fairProb >= 1) {
      throw new Error(`fairProb must be between 0 and 1, got ${fairProb}`);
    }
    this.fairProb = fairProb;
  }

  evaluate(snapshot: MarketSnapshot): TradeSignal {
    const [ok, reason] = this.passesQualityChecks(snapshot);
    if (!ok) return this.skip(reason);

    const yesPrice = snapshot.yesPrice;
    const edgeYes  = this.fairProb - yesPrice;
    const edgeNo   = yesPrice - this.fairProb;

    if (edgeYes >= this.config.minEdge) {
      const size = kellySize(this.fairProb, yesPrice, Action.BUY_YES, this.config);
      return {
        action:  Action.BUY_YES,
        tokenId: snapshot.yesTokenId,
        price:   snapshot.yesAsk ?? yesPrice,
        sizeUsd: size,
        edge:    edgeYes,
        reason:  `fair=${this.fairProb.toFixed(2)} market=${yesPrice.toFixed(2)} edge=${edgeYes.toFixed(3)}`,
      };
    }

    if (edgeNo >= this.config.minEdge) {
      const noPrice = snapshot.noPrice;
      const size = kellySize(this.fairProb, yesPrice, Action.BUY_NO, this.config);
      return {
        action:  Action.BUY_NO,
        tokenId: snapshot.noTokenId,
        price:   snapshot.noAsk ?? noPrice,
        sizeUsd: size,
        edge:    edgeNo,
        reason:  `fair=${this.fairProb.toFixed(2)} market=${yesPrice.toFixed(2)} edge=${edgeNo.toFixed(3)}`,
      };
    }

    return this.skip(
      `edge too small: yes_edge=${edgeYes.toFixed(3)} no_edge=${edgeNo.toFixed(3)} min=${this.config.minEdge.toFixed(3)}`,
    );
  }
}
